### DISCUSSION TITLE
Home Manager 2.0 consumption data?

### DISCUSSION THREAD

**Discussion by ziogref** [SBFspot/SBFspot#568] [General]
I have an SMA home manager 2. I know it uses Speedwire over my local network (Ethernet) to communicate with my Inverter (Also Ethernet) So my consumption data actually appears in in my inverter (as seen here https://imgur.com/uQZyEqk   ) from the screenshot you can see this is late at night and the inverter remains powered on 24/7 logging usage.

The home manager 2.0 doesn't have a web UI as much as I can tell, but the inverter can see import/export and generation and from that can work out self consumption. All that info is then available in the app and Sunny Portal website.

Is there a way for SBFspot to get that information to upload to PVportal?

side note: Devs you are fantastic and thanks for helping me with me issue prior.

![image](https://user-images.githubusercontent.com/1931158/189394966-4666b5fa-e31d-46ed-ab13-4a5c44d476c5.png)

**Comment by sbf-** [comment]
Related to [this](https://github.com/SBFspot/SBFspot/discussions/464#discussioncomment-2510958)

**Comment by sbf-** [comment]
As I explained to @g-mocken in [this discussion](https://github.com/SBFspot/SBFspot/discussions/464#discussioncomment-2510958), it might be possible recent inverters have some unexplored tables.

I don't have a Home Manager nor an Energy Meter, so my tables are empty, but if you want to test, here is how to proceed:
For the test we will (mis)use the `ArchiveDayData` function (this is the one used for the 5-min historical data)

Open the source file `ArchData.cpp` and locate the `ArchiveDayData` function (it's the very first one)
Find the line `writeLong(pcktBuf, 0x70000200);` (around line 93) - This is the command to get the 5-min data for a certain period.
The other known commands are `0x70100200` for Events (`ArchiveEventData` function) and `0x70200200` (`ArchiveMonthData` function) for Totals per day.

Now, I tried two extra functions `0x70300200` and `0x70400200` who do return valid data. The next two `0x70500200` and `0x70600200` do return only one record.

So for the test,  you only have to change the line to `writeLong(pcktBuf, 0x70300200);` or `writeLong(pcktBuf, 0x70400200);`
Rebuild the program with `make sqlite` and run it. Remember: we modified the `ArchiveDayData` function, so the -adxx is important here (xx = number of days to retrieve)

When this actually works I can include this in the program. I expect the first command to return 5-min consumption data and the second probably total consumption per day. 

I'm looking forward to the results and I'm sure this would be a really nice extension to SBFspot

> Important remark: since we modify the ArchiveDayData function, make sure to use -nosql or test with a new or copy of your database.

**Reply by ziogref** [reply]
Hmm this might be a little bit out of my depth, but I could backup my vdisk and give it a crack.

**Reply by ziogref** [reply]
Ok, so I think I did this correctly? I backed up my vdisk so I can restore and I also commented out the 2 jobs in there.

You were correct line 93 is what needed to be changed, so I changed that to `writeLong(pcktBuf, 0x70300200);` then ran `make sqlite` then ran `/usr/local/bin/sbfspot.3/SBFspot -v -finq -nocsv -ad01` (this command was the one I was unsure about, but I think is what you are after?)

This is the output of 0x70300200

```
SBFspot V3.9.6
Yet another tool to read power production of SMA solar inverters
(c) 2012-2022, SBF (https://github.com/SBFspot/SBFspot)
Compiled for Linux (LE) 64 bit with SQLite support
Commandline Args: -v -finq -nocsv -ad01
Reading config '/usr/local/bin/sbfspot.3/SBFspot.cfg'
Sat Sep 10 21:12:36 2022: INFO: Starting...
sunrise: 19:03
sunset : 06:34
Connecting to Local Network...
Initialising...
SUSyID: 125 - SessionID: 902316413
Device IP address: 10.1.1.230 from config
Inverter replied: 10.1.1.230 -> 409:3008366845
Logon OK
SUSyID: 409 - SN: 3008366845
Device Name:      SN: 3008366845
Device Class:     Solar Inverters
Device Type:      Sunny Boy 6.0
Software Version: 03.10.18.R
SUSyID: 409 - SN: 3008366845
Grid Power Out : 0W
Grid Power In  : 889W
SUSyID: 409 - SN: 3008366845
Device Status:      Ok
Packet status: 21
SUSyID: 409 - SN: 3008366845
Device Temperature: Information not available
SUSyID: 409 - SN: 3008366845
GridRelay Status:      Information not available
SUSyID: 409 - SN: 3008366845
Energy Production:
	EToday: 7.580kWh
	ETotal: 504.580kWh
	Operation Time: 236.50h
	Feed-In Time  : 222.52h
00464B01 28 Sat Sep 10 21:11:19 2022 'Grid voltage phase L1 against L2' NaN NaN NaN NaN
00464C01 28 Sat Sep 10 21:11:19 2022 'Grid voltage phase L2 against L3' NaN NaN NaN NaN
00464D01 28 Sat Sep 10 21:11:19 2022 'Grid voltage phase L3 against L1' NaN NaN NaN NaN
00464E01 28 Sat Sep 10 21:12:37 2022 '' 82 82 82 82
SUSyID: 409 - SN: 3008366845
DC Spot Data:
	MPPT 1 Pdc:   0.000kW - Udc:   0.00V - Idc:  0.000A
	MPPT 2 Pdc:   0.000kW - Udc:   0.00V - Idc:  0.000A
	Calculated Total Pdc:   0.000kW
AC Spot Data:
	Phase 1 Pac :   0.000kW - Uac:   0.00V - Iac:  0.000A
	Phase 2 Pac :   0.000kW - Uac:   0.00V - Iac:  0.000A
	Phase 3 Pac :   0.000kW - Uac:   0.00V - Iac:  0.000A
	Total Pac   :   0.000kW - Calculated Pac:   0.000kW
	Efficiency  :    0.00%
SUSyID: 409 - SN: 3008366845
Grid Freq. : 0.00Hz
SUSyID: 409 - SN: 3008366845
Current Inverter Time: 10/09/2022 21:12:37
Inverter Wake-Up Time: 10/09/2022 06:20:31
Inverter Sleep Time  : 10/09/2022 21:11:19
********************
* ArchiveDayData() *
********************
startTime: 10/09/2022 00:00:00
**********************
* ArchiveMonthData() *
**********************
startTime: 01/09/2022 12:00:00
Reading events: 2022-Sep-01
Sat Sep 10 21:12:37 2022: INFO: Done.
```

and this is the output of 0x70400200
```
SBFspot V3.9.6
Yet another tool to read power production of SMA solar inverters
(c) 2012-2022, SBF (https://github.com/SBFspot/SBFspot)
Compiled for Linux (LE) 64 bit with SQLite support
Commandline Args: -v -finq -nocsv -ad01
Reading config '/usr/local/bin/sbfspot.3/SBFspot.cfg'
Sat Sep 10 21:17:18 2022: INFO: Starting...
sunrise: 19:03
sunset : 06:34
Connecting to Local Network...
Initialising...
SUSyID: 125 - SessionID: 979583978
Device IP address: 10.1.1.230 from config
Inverter replied: 10.1.1.230 -> 409:3008366845
Logon OK
SUSyID: 409 - SN: 3008366845
Device Name:      SN: 3008366845
Device Class:     Solar Inverters
Device Type:      Sunny Boy 6.0
Software Version: 03.10.18.R
SUSyID: 409 - SN: 3008366845
Grid Power Out : 0W
Grid Power In  : 895W
SUSyID: 409 - SN: 3008366845
Device Status:      Ok
Packet status: 21
SUSyID: 409 - SN: 3008366845
Device Temperature: Information not available
SUSyID: 409 - SN: 3008366845
GridRelay Status:      Information not available
SUSyID: 409 - SN: 3008366845
Energy Production:
	EToday: 7.580kWh
	ETotal: 504.580kWh
	Operation Time: 236.50h
	Feed-In Time  : 222.52h
00464B01 28 Sat Sep 10 21:17:19 2022 'Grid voltage phase L1 against L2' NaN NaN NaN NaN
00464C01 28 Sat Sep 10 21:17:19 2022 'Grid voltage phase L2 against L3' NaN NaN NaN NaN
00464D01 28 Sat Sep 10 21:17:19 2022 'Grid voltage phase L3 against L1' NaN NaN NaN NaN
00464E01 28 Sat Sep 10 21:17:19 2022 '' 82 82 82 82
SUSyID: 409 - SN: 3008366845
DC Spot Data:
	MPPT 1 Pdc:   0.000kW - Udc:   0.00V - Idc:  0.000A
	MPPT 2 Pdc:   0.000kW - Udc:   0.00V - Idc:  0.000A
	Calculated Total Pdc:   0.000kW
AC Spot Data:
	Phase 1 Pac :   0.000kW - Uac:   0.00V - Iac:  0.000A
	Phase 2 Pac :   0.000kW - Uac:   0.00V - Iac:  0.000A
	Phase 3 Pac :   0.000kW - Uac:   0.00V - Iac:  0.000A
	Total Pac   :   0.000kW - Calculated Pac:   0.000kW
	Efficiency  :    0.00%
SUSyID: 409 - SN: 3008366845
Grid Freq. : 0.00Hz
SUSyID: 409 - SN: 3008366845
Current Inverter Time: 10/09/2022 21:17:19
Inverter Wake-Up Time: 10/09/2022 06:20:31
Inverter Sleep Time  : 10/09/2022 21:17:19
********************
* ArchiveDayData() *
********************
startTime: 10/09/2022 00:00:00
**********************
* ArchiveMonthData() *
**********************
startTime: 01/09/2022 12:00:00
Reading events: 2022-Sep-01
Sat Sep 10 21:17:19 2022: INFO: Done.
```

logging into my inverter (within a few minutes of running the above commands) this is a little after 9pm, so its been dark outside for sometime I have purchased from the grid 13.90 kWh and self consumed 5.2kWh giving a total of 19.1kWh used, exporting 2.4kWh with total generation of 7.6kWh (just to give you todays numbers to make things, maybe?, easier)

**Reply by sbf-** [reply]
Hmm, don't know if you copied the new compiled executable to `/usr/local/bin/sbfspot.3` - I didn't mention this step
It should be somewhere in subfolder `sqlite/bin`

Also, run SBFspot in full verbose mode (-v5) or even better with -d5 -v5
`/usr/local/bin/sbfspot.3/SBFspot -v5 -d5 -finq -nocsv -nosql -ad1 -am0 -ae0`

**Reply by ziogref** [reply]
> Hmm, don't know if you copied the new compiled executable to `/usr/local/bin/sbfspot.3` - I didn't mention this step It should be somewhere in subfolder `sqlite/bin`
> 
> Also, run SBFspot in full verbose mode (-v5) or even better with -d5 -v5 `/usr/local/bin/sbfspot.3/SBFspot -v5 -d5 -finq -nocsv -nosql -ad1 -am0 -ae0`

Ok the command I ran this time

inside ```/home/administrator/SBFspot/SBFspot/```
```nano ArchData.cpp```
edit line 93 as per example
```writeLong(pcktBuf, 0x70300200);```
```make sqlite```
```sudo cp /home/administrator/SBFspot/SBFspot/sqlite/bin/SBFspot /usr/local/bin/sbfspot.3/SBFspot```
```/usr/local/bin/sbfspot.3/SBFspot -v5 -d5 -finq -nocsv -nosql -ad1 -am0 -ae0 | tee /home/administrator/logs/[log].txt```

Logs on pastebin
[0x70300200](https://pastebin.com/yx3qAunP)
[0x70400200](https://pastebin.com/9V7A4AsN)
[0x70500200](https://pastebin.com/wZBBdVKu)
[0x70600200](https://pastebin.com/Wj1eyxJH)

[Today's stats screenshot from inverter webui](https://imgur.com/bNI8sKW)

**Reply by sbf-** [reply]
Well, this looks very promising! 0x70400200 shows us the imported energy as shown in this excel graph:

![image](https://user-images.githubusercontent.com/1931158/189846916-39f6a86a-3161-4b28-b30b-e345a8a33bd7.png)
![image](https://user-images.githubusercontent.com/1931158/189847374-4d28483f-a832-4226-bfef-2954d139eec3.png)

0x70300200 shows this (is this generated energy minus consumption?)
When you look at the max generated = 6000W and the max at below graph = 5500W (consumption around 500W ???)
![image](https://user-images.githubusercontent.com/1931158/189848126-ecd9ccb4-e1e0-4e76-9084-57e7f232c1cb.png)

**Reply by ziogref** [reply]
I have a home server and equipment that runs 24/7 so that would be the discrepancy on 0x70300200, which is about 500w

So I assume that 0x70400200 is import
and 0x70300200 is export

[generation]-[0x70300200]+[0x70400200] = usage

**Reply by alefhar** [reply]
Hi, since I have a Home Manager myself and am interested in getting the consumption data automatically I tried to replicate @ziogref results. They are indeed promising!

This is consumption data (usage, import, export, generation) that I get from a different tool:
![em_data](https://user-images.githubusercontent.com/191115/189889694-30fe2466-a412-42c5-8c54-8ed55420216c.png)

And this is the data I get from the inverter as per the method above:
![inverter_data](https://user-images.githubusercontent.com/191115/189889718-d3840caa-620f-4597-af85-c6c5bf949b90.png)

The inverter data has lower resolution but overall looks the same. Hope that helps.

I can help with testing, too.

**Reply by sbf-** [reply]
> [generation]-[export]+[import] = usage

So that's what you want to send to PVoutput? (Energy and Energy Used)

In the meantime I found out that 0x70500200 returns the daytotals for export and 0x70600200 for import

> The inverter data has lower resolution but overall looks the same

These are the averages over 5 minutes (HM sends its data every second)

**Reply by sbf-** [reply]
Q: does this work for multiple days (-ad2) ? Generation data is stored for +/- 62 days. Does this apply to import/export?

**Reply by alefhar** [reply]
> Q: does this work for multiple days (-ad2) ? Generation data is stored for +/- 62 days. Does this apply to import/export?

I've checked for 0x70300200 and 0x70400200 with -ad3 and got data for 2022-09-11 to 2022-09-13 as expected

**Reply by ziogref** [reply]
> So that's what you want to send to PVoutput? (Energy and Energy Used)

Probably just consumption (PV output calls it "Power used"). From that I assume PVoutput can work out import/export and give the daily cost of power. Which is kinda what I am after.

What I'm really after is pulling SBFspot data into my own tool, but that's a personal project further down the track, but if SBFspot could store Generation, Import, export, (and based on those values, consumption).

That way when I build a tool, I can hook into that data and use it in my own personal tool I'm am going to attempt to build.

**Reply by ziogref** [reply]
> This is consumption data (usage, import, export, generation) that I get from a different tool

What tool are you using? I want to start ripping some data out to start getting actual, this is how many KWh im self consuming so I can accurately calculate my savings. I only need accuracy to the hour (since I have TOU power)

**Reply by alefhar** [reply]
I'm using [SMA-EM](https://github.com/datenschuft/SMA-EM) that collects data from the inverter and the consumption data that is broadcast by the energy meter. I have the data simply stored in an InfluxDB and displayed with Grafana.
It should be possible to write a custom module to export consumption data to SBF's database for upload to PVOutput, but I haven't found the time for that, yet.

**Reply by ziogref** [reply]
> I'm using [SMA-EM](https://github.com/datenschuft/SMA-EM) that collects data from the inverter and the consumption data that is broadcast by the energy meter. I have the data simply stored in an InfluxDB and displayed with Grafana. It should be possible to write a custom module to export consumption data to SBF's database for upload to PVOutput, but I haven't found the time for that, yet.

Ok I have spent hours upon hours (literally from 5pm to midnight for a few nights now) and I cant figure this out. Could you please share your config for SMA-EM Maybe from that I can figure out how to get the data into influxDB.

I know nothing about either (SMA-EM and InfluxDB) and I am just hitting my head on a brick wall, it's so infuriating not being able to rip data into a usable format.

Basically I want exactly what you have in grafana

**Reply by alefhar** [reply]
@ziogref Sorry, I had no opportunity yet to get to my PC (and I was looking for a way to contact you privately). I'll post everything here as soon as I can

**Reply by alefhar** [reply]
@ziogref [This](https://pastebin.com/KzBPSX18) is my SMA-EM configuration. You have to fill in the fields with format `${identifier}` with the appropriate values.

**Reply by ziogref** [reply]
> @ziogref [This](https://pastebin.com/KzBPSX18) is my SMA-EM configuration. You have to fill in the fields with format `${identifier}` with the appropriate values.

Hey, im getting errors with the sma-daemon and it looks like im not the only one. would it be possible to get your copy of `/opt/smaemd`?

It looks like the daemon is calling an empty variable.

**Reply by alefhar** [reply]
@ziogref I remember SMA-EM causing some problems with my setup, too. But I just checked, the unit file points to `/opt/smaemd/sma-daemon.py` and this file does not have any changes.

**Reply by ziogref** [reply]
rats. I am trying home assistant again and seems to be able to pull in the data from the inverter and I am writing custom sensors to get the data I want, so hopefully I can get it working

**Comment by sbf-** [comment]
@ziogref
Q: the device with serial number xxxxxx0822 is that your Home Manager?

**Reply by ziogref** [reply]
> @ziogref Q: the device with serial number xxxxxx0822 is that your Home Manager?

Yes it is (Home Manager 2)

**Comment by g-mocken** [comment]
If I interpret the above discussion correctly, experiments on consumption data retrieval were successful. What else is missing to make the data show up in the consumption tables?
If there is more to test, let me know (sorry for being late to the party). I have an SMA Energymeter in place of the HomeManager, though.

**Comment by alefhar** [comment]
Hi @sbf- , is there something I could do to help with this enhancement?
